pkg/autoscaler/aggregation/bucketing2.go

```diff
@@ -0,0 +1,168 @@
+		granularity: granularity,
+		window:      window,
+	}
+}
+
+// timeToIndex converts time to an integer that can be used for modulo
+// operations to find the index in the bucket list.
+// bucketMutex needs to be held.
+func (t *TimedFloat64Buckets2) timeToIndex(tm time.Time) int {
+	return int(tm.Unix()) / int(t.granularity.Seconds())
+}
+
+func (t *TimedFloat64Buckets2) reset() {
+	for i := range t.buckets {
+		t.buckets[i] = 0
+	}
+}
+
+// Record adds a value with an associated time to the correct bucket.
+func (t *TimedFloat64Buckets2) Record(now time.Time, name string, value float64) {
+	bucketTime := now.Truncate(t.granularity)
+
+	t.bucketsMutex.Lock()
+	defer t.bucketsMutex.Unlock()
+
+	// I don't think this run in 2038 :-)
+	// NB: we need to divide by granularity, since it's a compressing mapping
+	// to buckets.
+	bucketIdx := t.timeToIndex(now) % len(t.buckets)
+
+	if t.lastWrite != bucketTime {
+		// This should not really happen, but is here for correctness.
+		if bucketTime.Sub(t.lastWrite) > t.window {
+			t.reset()
+		}
+		// Reset the value.
+		t.buckets[bucketIdx] = 0
+		// Update the last write time.
+		t.lastWrite = bucketTime
+	}
+	t.buckets[bucketIdx] += value
+}
+
+// isEmpty returns whether or not there are no values currently stored.
+// isEmpty requires t.bucketMux to be held and now truncated to the granularity.
+func (t *TimedFloat64Buckets2) isEmpty(now time.Time) bool {
+	return now.Sub(t.lastWrite) >= t.window
+}
+
+// ForEachBucket calls the given Accumulator function for each bucket.
+// Returns true if any data was recorded.
+func (t *TimedFloat64Buckets2) ForEachBucket(now time.Time, accs ...Accumulator) bool {
+	now = now.Truncate(t.granularity)
+	t.bucketsMutex.RLock()
+	defer t.bucketsMutex.RUnlock()
+
+	if t.isEmpty(now) {
+		return false
+	}
+
+	// So number of buckets we can process is len(buckets)-(now-lastWrite)/granularity.
+	// Since isEmpty returned false, we know this is at least 1 bucket.
+	nb := len(t.buckets) - int(now.Sub(t.lastWrite)/t.granularity)
+	tb := t.lastWrite // Always aligned with granularity.
+	si := t.timeToIndex(tb)
+	for i := 0; i < nb; i++ {
+		tix := si % len(t.buckets)
+		for _, acc := range accs {
+			acc(tb, t.buckets[tix])
+		}
+		si--
+		tb = tb.Add(-t.granularity)
+	}
+
+	return true
+}
+
+// RemoveOlderThan removes buckets older than the given time from the state.
+func (t *TimedFloat64Buckets2) RemoveOlderThan(time.Time) {
+	// RemoveOlderThan is a noop here.
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// ResizeWindow resizes the window. This is an O(N) operation,
+// and is not supposed to be executed very often.
+func (t *TimedFloat64Buckets2) ResizeWindow(w time.Duration) {
+	// Same window size, bail out.
+	if func() bool {
+		t.bucketsMutex.RLock()
+		defer t.bucketsMutex.RUnlock()
+		return w == t.window
+	}() {
+		return
+	}
+	nb := int(math.Ceil(float64(w) / float64(t.granularity)))
+	newb := make([]float64, nb)
```

Comment:
Can you please 🙏  not use shorthand variable names? 😂  `bucketCount`, `newBuckets` etc are so much easier to read than this.

Reply:
Go gods are throwing spears :-)

Reply:
updated most;